.ci/Jenkinsfile

```diff
@@ -0,0 +1,206 @@
+
+        runWithChecks = load "$jenkinsfileWorkspace/utils/runWithChecks.groovy"
+        expandDockerMatrix = load "$jenkinsfileWorkspace/utils/expandDockerMatrix.groovy"
+        prChangeset = load "$jenkinsfileWorkspace/utils/prChangeset.groovy"
+
+        pytorchDockerChanged = prChangeset("docker/pytorch/")
+    }
+}
+
+def dockerImagePostBuild(stagingImageTag) {
+    if (gitBranch == "main") {
+        // no need to run tests again
+        return
+    }
+    parallel([
+        "$stagingImageTag - all": { ->
+            build(
+                job: 'scratch/pytest',
+                parameters: [
+                    string(name: 'P_CLOUD', value: pCloud),
+                    string(name: 'P_GIT_REPO', value: gitUrl),
+                    string(name: 'P_GIT_COMMIT', value: gitCommit),
+                    string(name: 'P_DOCKER_IMAGE', value: stagingImageTag),
+                    string(name: 'P_CPU_LIMIT', value: "2"),
+                    string(name: 'P_MEM_LIMIT', value: "7Gi"),
+                    string(name: 'P_TIMEOUT', value: '1800'),
+                    string(name: 'PIP_EXTRA_DEPS', value: 'all'),
+                    string(name: 'PYTEST_COMMAND', value: './scripts/test.sh'),
+                    string(name: 'PYTEST_CLI_ARGS', value: '--test_duration all'),
+                ]
+            )
+        }
+    ])
+}
+
+stage('Build') {
+    def jobs = [:]
+    if (pytorchDockerChanged) {
+        jobs << expandDockerMatrix(
+            P_CLOUD: pCloud,
+            P_BUILD_MATRIX: './composer/pytorch_build_matrix.sh',
+            P_BUILD_MATRIX_GIT_REPO: 'https://github.com/mosaicml/testing.git',  // TODO RJPP_SCM_URL
+            P_BUILD_MATRIX_GIT_COMMIT: 'main', // TODO RJPP_BRANCH
+            P_DOCKERFILE: 'Dockerfile',
+            P_BUILD_CONTEXT: './docker/pytorch',
+            P_GIT_REPO: gitUrl,
+            P_GIT_COMMIT: gitCommit,
+            P_CPU_LIMIT: '4',
+            P_MEM_LIMIT: '15Gi',
+            P_TIMEOUT: '3600',
+            P_KANIKO_PUSH_FINAL: gitBranch == "main", // only push if we're on the main branch
+        ) { stagingImage -> dockerImagePostBuild(stagingImage) }
+    }
+    if (gitBranch != "main" && gitBranch != "dev") {
+        // if not on main or dev, run the pytest again.
+        jobs << [
+            'Lint': { -> 
+                runWithChecks(
+                    name: 'lint',
+                    title: 'Lint',
+                    summary: 'Static Analysis Checks',
+                ) {
+                    builds << build(
+                        job: 'scratch/lint',
+                        parameters: [
+                            string(name: 'P_CLOUD', value: pCloud),
+                            string(name: 'P_GIT_REPO', value: gitUrl),
+                            string(name: 'P_GIT_COMMIT', value: gitCommit),
+                            string(name: 'P_DOCKER_IMAGE', value: "mosaicml/pytorch:1.10.0_cpu-python3.9-ubuntu20.04"),
```

Comment:
shouldn't this image be calculated like it is above?

`"mosaicml/pytorch:${pytorchVersion}_${cudaVersion}-python${pythonVersion}-ubuntu20.04"`

Reply:
We only run linting on the latest version.